{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a683294-4a73-45c7-aad8-5bb9b1a9b088",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "import numpy as np\n",
    "from pandas import DataFrame\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import pysam\n",
    "from collections import Counter\n",
    "\n",
    "%matplotlib inline\n",
    "plt.rcParams['figure.figsize'] = [16.5, 5]\n",
    "plt.rcParams['font.size'] = 12"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f90f0e5-4070-429c-8d56-c383dfd76106",
   "metadata": {},
   "source": [
    "Define a function that extracts repeat allele sequences from each read. This function will return two lists containing sequences of each allele."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b68d528e-7986-493d-94cb-223754f0f3c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_seqs(path, trid):\n",
    "    bam = pysam.AlignmentFile(trgt_bam_path)\n",
    "    seqs_by_allele = ([], [])\n",
    "    for read in bam.fetch(until_eof=True):\n",
    "        allele = read.get_tag(\"AL\")\n",
    "        flank_lens = read.get_tag(\"FL\")\n",
    "        lf, rf = list(flank_lens)\n",
    "        seq = read.query_sequence[lf:-rf]\n",
    "        seqs_by_allele[allele].append(seq)\n",
    "    return seqs_by_allele"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57678ee6-b25e-4ef2-a053-b17be404ca8f",
   "metadata": {},
   "source": [
    "Specify a path to the BAM file we want to analyze. This must be the spanning reads BAM generated by TRGT."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49e32795-8623-4c80-b9a8-5d758b8c28af",
   "metadata": {},
   "outputs": [],
   "source": [
    "trgt_bam_path = \"input/trgt.bam\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48aa6a40-a755-4287-a289-2d4b2e9d7159",
   "metadata": {},
   "source": [
    "Let's use our function to extract repeat alleles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b95d3c92-4d3a-4e57-8e7a-f73fb49d6b35",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs_by_allele = get_seqs(trgt_bam_path, \"FMR1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42364f79-4423-4bc3-978c-08c20b9f83d7",
   "metadata": {},
   "source": [
    "Here are the sequences of the first allele:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74732fdc-beca-4574-a7ac-05f1ed229b70",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs_by_allele[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b674fbb9-2660-4e6f-8ce0-7b973e4298ef",
   "metadata": {},
   "source": [
    "And the sequences of the second allele:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e72cf452-f6e4-46cd-9bc6-7678b4816037",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs_by_allele[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "378e81ed-073f-4e42-81d7-c5a75ef9ad4d",
   "metadata": {},
   "source": [
    "Plot a histogram depicting the counts of reads with different length (stratified by allele)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8758f2da-a885-42b8-adaa-a708f67f5f5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bar_plot(values, ax, label):\n",
    "    counter = Counter(values)\n",
    "    xs = counter.keys()\n",
    "    ys = [counter[x] for x in xs]\n",
    "    ax.bar(xs, ys, label=label)\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(ncols=1, nrows=1, figsize=(5, 3))\n",
    "bar_plot([len(seq) for seq in seqs_by_allele[0]], ax, \"allele 1\")\n",
    "bar_plot([len(seq) for seq in seqs_by_allele[1]], ax, \"allele 2\")\n",
    "ax.set_xlabel(\"Repeat size (bps)\")\n",
    "ax.set_ylabel(\"Read count\")\n",
    "plt.legend();"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
